apis/supervisor/idp/v1alpha1/types_ldapidentityprovider.go.tmpl

```diff
@@ -0,0 +1,160 @@
+	// +kubebuilder:validation:MinLength=1
+	Host string `json:"host"`
+
+	// TLS contains the connection settings for how to establish the connection to the Host.
+	TLS *LDAPIdentityProviderTLSSpec `json:"tls,omitempty"`
+
+	// Bind contains the configuration for how to provide access credentials during an initial bind to the LDAP server
+	// to be allowed to perform searches and binds to validate a user's credentials during a user's authentication attempt.
+	Bind LDAPIdentityProviderBindSpec `json:"bind,omitempty"`
+
+	// UserSearch contains the configuration for searching for a user by name in the LDAP provider.
+	UserSearch LDAPIdentityProviderUserSearchSpec `json:"userSearch,omitempty"`
+
+	// DryRunAuthenticationUsername influences how the LDAPIdentityProvider's configuration is validated.
+	// When DryRunAuthenticationUsername is blank, the LDAPIdentityProvider will be validated by opening a connection
+	// to the LDAP server using the Host and TLS settings and also will bind using the Bind settings. The success
+	// or failure of the connect and bind will be reflected in the LDAPIdentityProvider's status conditions array.
+	// When DryRunAuthenticationUsername is not blank, the LDAPIdentityProvider will be validated by opening a
+	// connection to the LDAP server and performing a full dry run of authenticating as the end user with the username
+	// specified by DryRunAuthenticationUsername. The dry run will act as if the correct password were specified for
+	// that end user during the authentication. This will test all of the configuration options of the
+	// LDAPIdentityProvider. The success or failure of the authentication dry run will be reflected in the
+	// LDAPIdentityProvider's status conditions array, along with details of what username, UID, and group memberships
+	// were selected for the specified user. If the dry run fails, then that user would not be able to authenticate
+	// in a real authentication situation either, so the LDAPIdentityProvider's Status.Phase will be set to "Error".
+	// Therefore, the specified DryRunAuthenticationUsername must be a valid username of a real user who should be able
+	// to authenticate given all of the LDAPIdentityProvider's configuration. For example, if the UserSearch
+	// configuration were set up such that an end user should log in using their email address as their username, then
+	// the DryRunAuthenticationUsername should be the actual email address of a valid user who will be found in the LDAP
+	// server by the UserSearch criteria. Once you have used DryRunAuthenticationUsername to validate your
+	// LDAPIdentityProvider's configuration, you might choose to remove the DryRunAuthenticationUsername configuration
+	// if you are concerned that the user's LDAP account could change in the future, e.g. if the account could become
+	// disabled in the future.
+	DryRunAuthenticationUsername string `json:"dryRunAuthenticationUsername,omitempty"`
```

Comment:
I'd like to discuss this one more.

I'd expect that most/all users would want to use this field to get quick feedback on their settings, and that they would prefer that it be included in the CRD rather that have to learn about some other separate API or client which helps them perform a similar test. Although it's a long comment, the field is just a simple username string.

This API field feels super valuable, and is aligned with what users have already asked for in our user research. I don't see any risk in keeping it.

It could be moved to a deeper level of nesting like `authenticationDryRun.username` if we want to leave room for future dry run options, like specifying the expected username or an expected group name.

Reply:
tldr, please remove it.

1. This functionality is not part of this story
2. It can always be added later
3. This field is not config, it is a "test"
4. We have not reached agreement on the shape of this API


> I'd expect that most/all users would want to use this field to get quick feedback on their settings
> ...
> This API field feels super valuable, and is aligned with what users have already asked for in our user research.

I never once saw this requested in OpenShift.  At best it is a nice to have.

>  I don't see any risk in keeping it.

I do not see it as congruent with the rest of the API.

> and that they would prefer that it be included in the CRD rather that have to learn about some other separate API or client which helps them perform a similar test.

If this was a built-in Kube API, I would expect this to be a virtual sub resource `/logintest` that would take the username as input and return the Kube username and groups.  It would not change status or remove the IDP from the cache or anything like that.  Since we do not have the ability to create a sub resource, I would expect this to be a distinct resource.

> Although it's a long comment, the field is just a simple username string.

Even with the comment, I still did not understand what this field was doing until you explained it to me.

> It could be moved to a deeper level of nesting like `authenticationDryRun.username` if we want to leave room for future dry run options, like specifying the expected username or an expected group name.

I do not believe it belongs on this type at all.

Reply:
Removed in 36819989a30f8694cb2e70692898e845906d1f84